mlflow/tracking/fluent.py

```diff
@@ -1181,22 +1199,78 @@ def log_input(
+    dataset_input = _create_dataset_input(dataset, context, tags)
 
-    datasets = (
-        [DatasetInput(dataset=dataset._to_mlflow_entity(), tags=tags_to_log)] if dataset else None
+    MlflowClient().log_inputs(
+        run_id=run_id, datasets=dataset_input and [dataset_input], models=model and [model]
     )
 
-    MlflowClient().log_inputs(run_id=run_id, datasets=datasets, models=model and [model])
+
+def log_inputs(
+    datasets: list[Optional[Dataset]],
+    contexts: list[Optional[str]],
+    tags_list: list[Optional[dict[str, str]]],
+    models: list[Optional[LoggedModelInput]],
+) -> None:
+    """
+    Log a batch of datasets used in the current run.
+
+    The lists of `datasets`, `contexts`, `tags_list` must have the same length.
+    The entries in these lists can be ``None``, which represents empty value to the
+    corresponding input.
+
+    Args:
+        datasets: List of :py:class:`mlflow.data.dataset.Dataset` object to be logged.
+        contexts: List of context in which the dataset is used. For example: "training", "testing".
+            This will be set as an input tag with key `mlflow.data.context`.
+        tags_list: List of tags to be associated with the dataset. Dictionary of tag_key -> tag_value.
+        models: List of :py:class:`mlflow.entities.LoggedModelInput` instance to log as as input to the
+            run.
+
+    .. code-block:: python
+        :test:
+        :caption: Example
+
+        import numpy as np
+        import mlflow
+
+        array = np.asarray([[1, 2, 3], [4, 5, 6], [7, 8, 9]])
+        dataset = mlflow.data.from_numpy(array, source="data.csv")
+
+        array2 = np.asarray([[-1, 2, 3], [-4, 5, 6]])
+        dataset2 = mlflow.data.from_numpy(array2, source="data2.csv")
+
+        # Log 2 input datasets used for training and test,
+        # the training dataset has no tag, but has `model1` reference,
+        # the test dataset has tags `{"my_tag": "tag_value"}`, but has no model reference.
+        with mlflow.start_run():
+            mlflow.log_inputs(
+                [dataset, dataset2],
+                contexts=["training", "test"],
+                tags_list=[None, {"my_tag": "tag_value"},
+                models=[model1]
+            )
+    """
+    run_id = _get_or_start_run().info.run_id
+
+    if not (
+        datasets is not None
+        and contexts is not None
+        and tags_list is not None
+        and len(datasets) == len(contexts) == len(tags_list)
+    ):
+        raise MlflowException(
+            "`mlflow.log_inputs` requires `datasets`, `contexts`, `tags_list` to be "
+            "non-empty list and have the same length."
+        )
+    dataset_inputs = []
+    for dataset, context, tags in zip(datasets, contexts, tags_list):
+        dataset_inputs.append(_create_dataset_input(dataset, context, tags))
+
+    MlflowClient().log_inputs(
+        run_id=run_id, datasets=dataset_inputs, models=models
```

Comment:
Note: the `models` list length does not need to equal with dataset list length. They should be not relevant .

Reply:
If they don't equal one another could the mapping be indexed incorrectly?

Reply:
datasets and models does not have 1:1 mapping relationship , confirmed with @dbczumar